docs/core-concepts/params.md

```diff
@@ -0,0 +1,168 @@
+# Params
+Params are used to define the dynamic parts of a route. Params can be used in the `path` and `query` properties when defining a route.
+
+## Param Names
+Param names are defined using square brackets. What is inside the brackets is the name of the param and is used when accessing the param value and when providing the param value when navigating to the route.
+
+```ts {3}
+const events = createRoute({
+  name: 'events',
+  path: '/events/[year]/[month]',
+})
+```
+
+:::warning
+Param names must be unique across the `path` and `query` properties and parent routes.
+:::
+
+## Optional Params
+Params can be made optional by adding a `?` before the name. The `year` param is not optional and will have the type `string` when being accessed. But the `month` param is optional and will have the type `string | undefined` when being accessed.
+
+```ts {3}
+const events = createRoute({
+  name: 'events',
+  path: '/events/[year]/[?month]',
+})
+```
+
+## Param Types
+By default all params are strings. But you can define a specific type for a param by using the `path` or `query` utilities. Now the `year` param will have the type `number` when being accessed.
+
+```ts {1,5-7}
+import { path } from '@kitbag/router'
+
+const events = createRoute({
+  name: 'events',
+  path: path('/events/[year]/[?month]', {
+    year: Number,
+  }),
+})
+```
+
+:::info Route Matching
+When using a param type, if the url cannot be parsed as the specified type it will not be matched. For example if the url was `/events/two-thousand-and-twenty-four/september` it would not be matched because `two-thousand-and-twenty-four` is not a number.
```

Comment:
```suggestion
When using a param type, if the url cannot be parsed as the specified type it will not be matched. For example if the url was `/events/two-thousand-and-twenty-four/september` it would not be matched because `two-thousand-and-twenty-four` is not a number. See [route matching](/advanced-concepts/route-matching) for more information.
```